state/protocol/badger/mutator.go

```diff
@@ -526,99 +527,55 @@ func (m *FollowerState) Finalize(ctx context.Context, blockID flow.Identifier) e
+			return fmt.Errorf("could not check whether finalized block triggers epoch fallback: %w", err)
 		}
-		for _, event := range result.ServiceEvents {
-			switch ev := event.Event.(type) {
-			case *flow.EpochSetup:
-				// update current epoch phase
-				events = append(events, func() { m.metrics.CurrentEpochPhase(flow.EpochPhaseSetup) })
-				// track epoch phase transition (staking->setup)
-				events = append(events, func() { m.consumer.EpochSetupPhaseStarted(ev.Counter-1, header) })
-			case *flow.EpochCommit:
-				// update current epoch phase
-				events = append(events, func() { m.metrics.CurrentEpochPhase(flow.EpochPhaseCommitted) })
-				// track epoch phase transition (setup->committed)
-				events = append(events, func() { m.consumer.EpochCommittedPhaseStarted(ev.Counter-1, header) })
-				// track final view of committed epoch
-				nextEpochSetup, err := m.epoch.setups.ByID(epochStatus.NextEpoch.SetupID)
-				if err != nil {
-					return fmt.Errorf("could not retrieve setup event for next epoch: %w", err)
-				}
-				events = append(events, func() { m.metrics.CommittedEpochFinalView(nextEpochSetup.FinalView) })
-			default:
-				return fmt.Errorf("invalid service event type in payload (%T)", event)
-			}
+		if epochFallbackTriggered {
+			// emit the protocol event only the first time epoch fallback is triggered
+			events = append(events, m.consumer.EpochEmergencyFallbackTriggered)
 		}
 	}
 
-	// FOURTH: preparing Epoch-Change service notifications and metrics updates.
-	// Convention:
-	// Service notifications and updating metrics happen when we finalize the _first_
-	// block of the new Epoch (same convention as for Epoch-Phase-Changes)
-	// Approach: We retrieve the parent block's epoch information. If this block's view
-	// exceeds the final view of its parent's current epoch, this block begins the next epoch.
-	parentBlocksEpoch := m.AtBlockID(header.ParentID).Epochs().Current()
-	parentEpochFinalView, err := parentBlocksEpoch.FinalView()
-	if err != nil {
-		return fmt.Errorf("could not get parent epoch final view: %w", err)
-	}
-
-	// When this block's view exceeds the parent epoch's final view, this block
-	// represents the first block of the next epoch. Therefore we update metrics
-	// related to the epoch transition here.
-	//
-	// We skip updating these metrics when EECC has been triggered
-	if header.View > parentEpochFinalView && !epochFallbackTriggered {
-		events = append(events, func() { m.consumer.EpochTransition(currentEpochSetup.Counter, header) })
+	// Determine metric updates and protocol events related to epoch phase
+	// changes and epoch transitions.
+	// If epoch emergency fallback is triggered, the current epoch continues until
+	// the next spork - so skip these updates.
+	if !epochFallbackTriggered {
+		epochPhaseMetrics, epochPhaseEvents, err := m.epochPhaseMetricsAndEventsOnBlockFinalized(header, epochStatus)
```

Comment:
Refactored epoch-related metrics, events, and state changes into:
* `epochPhaseMetricsAndEventsOnBlockFinalized`
* `epochTransitionMetricsAndEventsOnBlockFinalized`
* `epochFallbackTriggeredByFinalizedBlock`